api/docs/mapstruct.md

```diff
@@ -0,0 +1,608 @@
+public class EmployeeDbEntity {
+
+  private String firstName;
+  private String lastName;
+  private int departmentCode; // corresponds to enum
+  private double hourlyRate;
+  private String address;
+
+  public EmployeeDbEntity(String firstName, String lastName, int departmentCode, double hourlyRate,
+      String address) {
+    this.firstName = firstName;
+    this.lastName = lastName;
+    this.departmentCode = departmentCode;
+    this.hourlyRate = hourlyRate;
+    this.address = address;
+  }
+
+  public String getFirstName() {
+    return firstName;
+  }
+  public void setFirstName(String firstName) {
+    this.firstName = firstName;
+  }
+
+  public int getDepartmentCode() {
+    return departmentCode;
+  }
+  public void setDepartmentCode(int departmentCode) {
+    this.departmentCode = departmentCode;
+  }
+
+  public double getHourlyRate() {
+    return hourlyRate;
+  }
+  public void setHourlyRate(double hourlyRate) {
+    this.hourlyRate = hourlyRate;
+  }
+
+  public String getLastName() {
+    return lastName;
+  }
+  public void setLastName(String lastName) {
+    this.lastName = lastName;
+  }
+
+  public String getAddress() {
+    return address;
+  }
+  public void setAddress(String address) {
+    this.address = address;
+  }
+}
+```
+
+A naive first cut at a mapping function might look like this:
+```java
+import org.mapstruct.Mapper;
+
+@Mapper
+public interface EmployeeMapper {
+
+  EmployeeModel toModel(EmployeeDbEntity employeeDbEntity);
+}
+```
+
+Using `./gradlew compileJava`, I see that this compiles, but gives warnings
+```
+examples/EmployeeMapper.java:8: warning: Unmapped target properties: "name, department, salary".
+  EmployeeModel toModel(EmployeeDbEntity employeeDbEntity);
+                ^
+```
+
+So the good news is I've successfully generated a mapper. The bad news is that it's only one for
+four properties.
+```java
+@Generated(
+    value = "org.mapstruct.ap.MappingProcessor",
+    date = "2020-05-02T18:10:18-0400",
+    comments = "version: 1.3.1.Final, compiler: javac, environment: Java 1.8.0_221 (Oracle Corporation)"
+)
+public class EmployeeMapperImpl implements EmployeeMapper {
+
+    @Override
+    public EmployeeModel toModel(EmployeeDbEntity employeeDbEntity) {
+        if ( employeeDbEntity == null ) {
+            return null;
+        }
+
+        EmployeeModel employeeModel = new EmployeeModel();
+
+        employeeModel.setAddress( employeeDbEntity.getAddress() );
+
+        return employeeModel;
+    }
+}
+```
+
+To fix the name mapping we need to add a method to build a full name from first & last names. Let's pretend our
+system didn't already have a good one, and add it directly in the mapper itself for brevity. To do
+this, we need to create a `default` method on the interface, which is a relatively new facility in
+Java to make life easier in this kind of situation.
+
+Our default method is itself another mapper, though you want to be careful when the source type of
```

Comment:
For what reason do you want to be careful? What does being careful or not being careful look like? What happens when you are not careful?